ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Stages/PipeChainTypeHintStage.fs

```diff
@@ -0,0 +1,153 @@
+
+    let [<Literal>] opName = "PipeChainHighlightingProcess"
+
+    /// Formats a type parameter layout.
+    /// Removes the "'T1 is " prefix from the layout string.
+    let formatTypeParamLayout (layout : Layout) =
+        let typeParamStr = showL layout
+        let prefixToRemove = "'T1 is "
+        if typeParamStr.StartsWith(prefixToRemove) then
+            typeParamStr.Substring(prefixToRemove.Length)
+        else
+            null
+
+    let adornExprs logKey checkResults (exprs : (FSharpIdentifierToken * ITreeNode)[]) =
+        use _swc = logger.StopwatchCookie(sprintf "Adorning %s expressions" logKey, sprintf "exprCount=%d sourceFile=%s" exprs.Length daemonProcess.SourceFile.Name)
+        let highlightingConsumer = FilteringHighlightingConsumer(daemonProcess.SourceFile, fsFile, settings)
+
+        for token, exprToAdorn in exprs do
+            if daemonProcess.InterruptFlag then raise <| OperationCanceledException()
+
+            let (FSharpToolTipText layouts) = FSharpIdentifierTooltipProvider.GetFSharpToolTipText(checkResults, token)
+
+            // The |> operator should have one overload and two type parameters
+            match layouts with
+            | [ FSharpStructuredToolTipElement.Group [ { TypeMapping = [ argumentType; _ ] } ] ] ->
+                let returnTypeStr = formatTypeParamLayout argumentType
+                if returnTypeStr = null then () else
+
+                // Use EndOffsetRange to ensure the adornment appears at the end of multi-line expressions
+                let range = exprToAdorn.GetNavigationRange().EndOffsetRange()
+
+                highlightingConsumer.AddHighlighting(TypeHintHighlighting(returnTypeStr, range))
+            | _ -> ()
+
+        highlightingConsumer.Highlightings
+
+    override x.Execute(committer) =
+        let sameLinePipeHints =
+            if settings.GetValue(fun (key: FSharpTypeHintOptions) -> key.HideSameLine) then
+                SameLinePipeHints.Hide
+            else
+                SameLinePipeHints.Show
+
+        match fsFile.GetParseAndCheckResults(true, opName) with
+        | None -> ()
+        | Some results ->
+
+        let consumer = List()
+        fsFile.Accept(PipeOperatorVisitor(fsFile, results.CheckResults, sameLinePipeHints), consumer)
+        let allHighlightings = Array.ofSeq consumer
+
+        // Visible range may be larger than document range by 1 char
+        // Intersect them to ensure commit doesn't throw
+        let documentRange = daemonProcess.Document.GetDocumentRange()
+        let visibleRange = daemonProcess.VisibleRange.Intersect(&documentRange)
+
+        let remainingHighlightings =
+            if visibleRange.IsValid() then
+                // Partition the expressions to adorn by whether they're visible in the viewport or not
+                let visible, notVisible =
+                    allHighlightings
+                    |> Array.partition (fun (token, exprToAdorn) ->
```

Comment:
`token` is unused too. It wasn't reported due to a bug in EAP builds in FCS assembly reader failure which could prevent doing some analysis, and it's fixed in the release.